src/fixate/core/switching.py

```diff
@@ -0,0 +1,792 @@
+        self._dispatch_pin_state(collated.final)
+
+    def _dispatch_pin_state(self, new_state: PinSetState) -> None:
+        # check all pins actually have an address handler to send to
+        if unknown_pins := (new_state.on | new_state.off) - self._all_pins:
+            raise ValueError(f"Can't switch unknown pin(s) {', '.join(unknown_pins)}.")
+
+        new_active_pins = (self._active_pins | new_state.on) - new_state.off
+        if new_active_pins != self._active_pins:
+            self._active_pins = new_active_pins
+            for pin_set, handler in self._handler_pin_sets:
+                # Note that we might send an empty set here. We need to do that
+                # so if there are pins to clear, they get cleared. This might
+                # end up in redundant handler updates, but unless we track active_pins
+                # per-handler I don't think we can avoid that.
+                handler.set_pins(pin_set & self._active_pins)
+
+    def active_pins(self) -> frozenset[Pin]:
+        return frozenset(self._active_pins)
+
+    def reset(self) -> None:
+        """
+        Sets all pins to be inactive.
+
+        Note: this does not change the state of any VirtualMux's, so it is
+        possible the state of each VirtualMux and its related pins will not
+        be in sync.
+        """
+        self._dispatch_pin_state(PinSetState(off=self._all_pins))
+
+    def update_input(self) -> None:
+        """
+        Currently not implemented. A few jigs implement digital input, but not many. It is the intention
+        to implement this eventually, but for now, the old jig_mapping.py version can be used.
+        """
+        raise NotImplementedError
+
+
+class MuxGroup:
+    """
+    Group multiple VirtualMux's, for use in a single Jig Driver.
+
+    If a test script, it is expected that MuxGroup will be subclassed, with attributes
+    being each required VirtualMux subclass. This can be done using a dataclass:
+
+    @dataclass
+    class JigMuxGroup(MuxGroup):
+        mux_one: MuxOne = field(default_factory=MuxOne)
+        mux_two: MuxTwo = field(default_factory=MuxTwo)
+    """
+
+    def get_multiplexers(self) -> list[VirtualMux]:
+        return [attr for attr in self.__dict__.values() if isinstance(attr, VirtualMux)]
+
+    def reset(self) -> None:
+        mux_list = self.get_multiplexers()
+        if len(mux_list) == 0:
+            return
+
+        for mux in mux_list[:-1]:
+            mux.reset(trigger_update=False)
+        mux_list[-1].reset(trigger_update=True)
+
+    def active_signals(self) -> list[str]:
+        return [str(mux) for mux in self.get_multiplexers()]
+
+
+JigSpecificMuxGroup = TypeVar("JigSpecificMuxGroup", bound=MuxGroup)
+
+
+class JigDriver(Generic[JigSpecificMuxGroup]):
+    """
+    Combine multiple VirtualMux's and multiple AddressHandler's.
+
+    The jig driver joins muxes to handlers by matching up pin definitions.
+    """
+
+    def __init__(
+        self,
```

Comment:
Do we plan on adding a runtime check to make sure all the pins in the muxgroups are implemented in the handlers?

Reply:
Could add an assert in the loop that registers `add_update` with each mux.
`assert set(mux.pin_list) <= self.virtual_map._all_pins, "mux pins not in map"`

Reply:
Or raise an exception?

Reply:
There is a runtime check that the pin from a virtual mux exists when it gets dispatched to the address handler https://github.com/PyFixate/Fixate/pull/184/files/f464c4e3c5841bb9ff98fae8ed6f937e31f4cec6#diff-9446774e4fed19c7eec4a6f2aec5f67a9ab22433bdbbea3ae51f9359bfa3ee79R608

> For the converse where there are more pins in the handlers than there are in muxes, does it really matter if the pins supplied are sufficient?

I think there is good reason to allow this. It is reasonable to imagine defining "spare" pins. You essentially have to in order for the address handler definition to work.

However I have been thinking of a linting/reporting tool. This could be run manually or part of the script zip process and could error on pins in virtual muxs missing from handlers, warn on unused address handler pins etc

Reply:
Was that link above meant to land on this bit of code?
![image](https://github.com/PyFixate/Fixate/assets/28747582/1a5733a0-2046-4c4f-9e8c-1a91a0cf0c75)

Reply:
Also, agree RE spare pins.

Reply:
> Was that link above meant to land on this bit of code?
Yep